### PR TITLE
Meet smiley

### PR DESCRIPTION
**Pull request by AljoschaMeyer**
This module ended up being fairly uninteresting audiowise, but the theming is top-notch!

### CODE REVIEW

**Comment by AljoschaMeyer** [comment]
And here is another one I wanted to implement since Monday: clippy! Clippy clips amplitudes to a threshold.

But that's not all, clippy also has an alignment, which is `lawful good` by default. If you set it to chaotic rather than lawful, it does the opposite of clipping: the absolute values of the amplitudes will be at least the threshold, not at most. And/or if you set it to evil rather than good, it eats samples rather than fixing them.

This ends up being a superset of what hungry can do, I hope that isn't a problem.

**Comment by AljoschaMeyer** [comment]
Oh no, I cannot stop. This is *extremely* addictive.

**Comment by AljoschaMeyer** [comment]
Interpolating toward random values turned out to be surprisingly boring, you just get some white noise in the background. Which is completely obvious in hindsight, but I had somehow hoped for distortion.

**Comment by AljoschaMeyer** [comment]
And lastly for today, *speedy* can adjust the speed by duplicating/dropping samples as necessary. Conceptually rather uninteresting, but auditively very rewarding =)

**Comment by AljoschaMeyer** [comment]
I should sleep now...

File: src/clippy.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+use earwig::utils::next_sample;
+
+fn main() {
+    // Parse args or set defaults
+    let args: Vec<String> = env::args().collect();
+
+    // Set the amplitude threshold for clippy to interact with samples.
+    let threshold: f64 = match args.get(1) {
+        Some(ms) => ms.parse().unwrap_or(0.3),
```
Comment (sandreae) [inline src/clippy.rs:11]: Guess this should be something other than `ms`?

File: src/extremely.rs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+use earwig::utils::{next_sample, lerp};
+
+fn main() {
+    // Parse args or set defaults
+    let args: Vec<String> = env::args().collect();
+
+    // How strongly to interpolate samples toward +/- 1.
+    let t: f64 = match args.get(1) {
+        Some(ms) => ms.parse().unwrap_or(1.0),
```
Comment (sandreae) [inline src/extremely.rs:11]: same here (not `ms`)

File: src/clippy.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        // Oops, we forgot how clipping is supposed to work.
+        if !lawful && sample.abs() < threshold {
+            if good {
+                // We can "fix" this sample!
+                sample = threshold * sample.signum();
+            } else {
+                // This sample has no right to pass through *evil laugh*!
+                continue
+            }
+        }
```
Comment (sandreae) [inline src/clippy.rs:61]: Love Love Love This! I want to frame this code snippet.

File: src/randomly.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    return state;
+}
+
+fn main() {
+    // Parse args or set defaults
+    let args: Vec<String> = env::args().collect();
+
+    // How strongly to interpolate samples toward chaos.
+    let t: f64 = match args.get(1) {
+        Some(ms) => ms.parse().unwrap_or(0.5),
```
Comment (sandreae) [inline src/randomly.rs:19]: here too

File: src/randomly.rs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+use earwig::utils::{next_sample, lerp};
+
+// https://en.wikipedia.org/wiki/Xorshift
+fn xorshift(mut state: u32) -> u32 {
+    state ^= state << 13;
+    state ^= state >> 17;
+    state ^= state << 5;
+    return state;
+}
```
Comment (sandreae) [inline src/randomly.rs:11]: :heart:

**Review by sandreae — APPROVED** [review]
Oh god, this is all so good. Any one of these modules could be a stand alone instrument I would happily play for hours. 

Thanks for this awesome PR, it feels like [insert seasonal gift giving tradition] came early.

(I just had a tiny copy-paste nit-pick which I can fix post merge)

**Pull request merged**